### PR TITLE
Bug 1690259 - Keep historical data

### PR DESCRIPTION
**Pull request by beatrice-acasandrei**

### CODE REVIEW

File: treeherder/perf/models.py [hunk 1]
```diff
@@ -100,6 +101,30 @@ def has_performance_data(self):
             signature_id=self.id,
         ).exists()
 
+    def has_data_with_historical_value(self):
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:104]: This starts to look like a big method & isn't easy to follow. Let's try & keep it under 10 lines of code.

File: treeherder/perf/models.py [hunk 2]
```diff
@@ -100,6 +128,26 @@ def has_performance_data(self):
...
+            perf_data = list(
+                PerformanceDatum.objects.filter(
+                    repository_id=self.repository_id,  # leverages (repository, signature) compound index
+                    signature_id=self.id,
+                ).order_by("push_timestamp")
+            )
+
+            most_recent_perf_data = perf_data[-1].push_timestamp
+
+            if data_was_kept_for_one_year(most_recent_perf_data):
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:143]: We don't need this check. There's a separate strategy that removes data older than one year.

File: treeherder/perf/models.py [hunk 3]
```diff
@@ -26,6 +27,33 @@ def __str__(self):
         return self.name
 
 
+def has_one_month_worth_of_data(perf_data):
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:30]: These helper functions should be used as private/protected methods inside the main class.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

File: treeherder/perf/models.py [hunk 4]
```diff
@@ -94,12 +94,45 @@ def _get_alert_change_type(alert_change_type_input):
                 return idx
         return None
 
+    @staticmethod
+    def _has_one_month_worth_of_data(perf_data):
```
Comment (airimovici) [inline treeherder/perf/models.py:98]: maybe rename this to `_has_enough_data`

File: treeherder/perf/models.py [hunk 5]
```diff
@@ -94,12 +94,43 @@ def _get_alert_change_type(alert_change_type_input):
...
 
+    @staticmethod
+    def _has_enough_data_points(perf_data):
+        start_date = perf_data[0].push_timestamp
+        end_date = perf_data[-1].push_timestamp
+
+        perf_data_count = len(perf_data)
+
+        num_months = (
+            (end_date.year - start_date.year) * 12 + (end_date.month - start_date.month) + 1
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:105]: Why have we incremented this by one?

File: tests/model/cycle_data/test_perfherder_cycling.py [hunk 6]
```diff
@@ -595,6 +595,136 @@ def test_stalled_data_removal(
...
+    test_perf_signature_2,
+    test_perf_data,
+):
+    nr_months = 8
+    test_repository.name = 'autoland'
+    test_repository.save()
+    four_months_ago_timestamp = datetime.now() - timedelta(days=120)
+    timestamp = datetime.now() - timedelta(days=((nr_months * 30) + 120 + 30))
+    test_perf_signature = test_perf_signature_2
+    test_perf_signature.last_updated = four_months_ago_timestamp
```
Comment (ionutgoldan) [inline tests/model/cycle_data/test_perfherder_cycling.py:701]: I see we prepare this same signature in all 3 tests. We could consider DRYing it.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

File: tests/model/cycle_data/test_perfherder_cycling.py [hunk 7]
```diff
@@ -595,6 +595,136 @@ def test_stalled_data_removal(
...
+            job=None,
+            signature=test_perf_signature,
+            push_timestamp=timestamp,
+            value=1.0,
+        )
+        perf_data.append(data)
+    call_command('cycle_data', 'from:perfherder')
+
+    assert test_perf_signature not in PerformanceSignature.objects.all()
+    for data_point in perf_data:
```
Comment (ionutgoldan) [inline tests/model/cycle_data/test_perfherder_cycling.py:724]: This assert should be optimized, to run faster.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

File: tests/conftest.py [hunk 8]
```diff
@@ -528,6 +528,23 @@ def test_perf_signature_2(test_perf_signature):
     )
 
 
+@pytest.fixture
+def test_perf_signature_3(test_perf_signature):
```
Comment (ionutgoldan) [inline tests/conftest.py:532]: Could you give this fixture a more meaningful name? The "3" suffix is totally opaque.

File: tests/model/cycle_data/test_perfherder_cycling.py [hunk 9]
```diff
@@ -595,6 +595,155 @@ def test_stalled_data_removal(
...
+            job=None,
+            signature=perf_signature,
+            push_timestamp=timestamp,
+            value=1.0,
+        )
+        perf_data.append(data)
+
+    call_command('cycle_data', 'from:perfherder')
+
+    assert perf_signature in PerformanceSignature.objects.all()
```
Comment (ionutgoldan) [inline tests/model/cycle_data/test_perfherder_cycling.py:638]: I still think we could optimize these asserts.

File: treeherder/perf/models.py [hunk 10]
```diff
@@ -94,12 +94,40 @@ def _get_alert_change_type(alert_change_type_input):
...
+    def has_data_with_historical_value(self):
+        repositories = ['autoland', 'mozilla-central']
+        if self.repository.name in repositories:
+            perf_data = list(
+                PerformanceDatum.objects.filter(
+                    repository_id=self.repository_id,  # leverages (repository, signature) compound index
+                    signature_id=self.id,
+                ).order_by("push_timestamp")
+            )
+            if len(perf_data) >= 2 and self._has_enough_data_points(perf_data):
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:127]: I think we can move this `len()` check directly in the protected method.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]
I forgot to tell you: we also need to update the online docs [here](https://treeherder.readthedocs.io/data_cycling.html), to reflect the changes.

**Review by ionutgoldan — APPROVED** [review]

**Pull request merged**